### PR TITLE
Update geos.py to support macOS Big Sur

### PR DESCRIPTION
**Pull request by niklasarnitz**
In the latest macOS (Big Sur), apple changed the way libraries are found and linked.
In the current version of shapely, nothing works because of this:
```
Traceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 968, in _find_and_load
  File "<frozen importlib._bootstrap>", line 957, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 697, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "/Users/ultimaker/build/4.6/build/inst/lib/python3.5/site-packages/cura/CuraApplication.py", line 51, in <module>
  File "/Users/ultimaker/build/4.6/build/inst/lib/python3.5/site-packages/cura/Arranging/Arrange.py", line 7, in <module>
  File "/Users/ultimaker/build/4.6/build/inst/lib/python3.5/site-packages/UM/Math/Polygon.py", line 10, in <module>
  File "/Users/ultimaker/build/4.6/build/inst/lib/python3.5/site-packages/UM/Math/ShapelyUtil.py", line 6, in <module>
  File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/geometry/__init__.py", line 4, in <module>
  File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/geometry/base.py", line 17, in <module>
  File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/coords.py", line 8, in <module>
  File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/geos.py", line 113, in <module>
  File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/geos.py", line 56, in load_dll
OSError: Could not find lib c or load any of its variants [].
```

The change in geos is only because it fixes the error in Cura (see this Issue: [#7970] (https://github.com/Ultimaker/Cura/issues/7970) over at Cura) and makes it work on Big Sur.
I thought I would create a PR so you'll see it :)

### CODE REVIEW

**Comment by Ghostkeeper** [comment]
This seems to revert a number of changes that were recently made to Shapely. Maybe we need to filter down what needs to change, because the version in Cura was not completely up to date with master here.

File: shapely/geos.py [hunk 1]
```diff
@@ -21,12 +21,19 @@
 # Add message handler to this module's logger
 LOG = logging.getLogger(__name__)
 
+if sys.version_info[0] >= 3:
```
Comment (snorfalorpagus) [inline shapely/geos.py:24]: We aren't planning to support Python 2.x in Shapely 1.8 so these changes aren't needed (and aren't wanted).

File: shapely/geos.py [hunk 2]
```diff
@@ -74,7 +77,7 @@ def exists_conda_env():
         if len(geos_pyinstaller_so) == 1:
             _lgeos = CDLL(geos_pyinstaller_so[0])
             LOG.debug("Found GEOS DLL: %r, using it.", _lgeos)
-    elif exists_conda_env():
+    elif os.getenv('CONDA_PREFIX', ''):
```
Comment (snorfalorpagus) [inline shapely/geos.py:80]: I don't know conda well enough to comment on this. Are there any docs/examples on why this is the best way? It looks sensible at least.

File: shapely/geos.py [hunk 3]
```diff
@@ -869,23 +875,9 @@ class LGEOS360(LGEOS350):
 
     def __init__(self, dll):
         super(LGEOS360, self).__init__(dll)
-        self.methods['minimum_clearance'] = self.GEOSMinimumClearance
-
-
-class LGEOS380(LGEOS360):
```
Comment (snorfalorpagus) [inline shapely/geos.py:875]: What's the reason for these changes? Was this commited by mistake?

Comment (Ghostkeeper) [reply shapely/geos.py:875]: This was recently added to Shapely here: https://github.com/Toblerity/Shapely/commit/087e53d3a34e79211f50b34740c938d2690d3e81

So yes, I believe that Niklasarnitz mistakenly copied over a (modified) older version of Shapely on top of this file, undoing all recent changes. Basically a mistake in merging.

Comment (niklasarnitz) [reply shapely/geos.py:875]: Oh yes, I am sorry. I kinda just hacked it together myself to get Cura working on macOS Big Sur. I thought I should post my changes here, bc they might help.

Comment (gunterthomas) [reply shapely/geos.py:875]: I guess it would be cool if you revert these changes to the way they should be - so maybe this Pull-Request would be merged soon and we all could start building Cura for MacOS Big Sur.

File: shapely/geos.py (hunk 2 above)
Comment (jorisvandenbossche) [reply shapely/geos.py:80]: This was changed from the CONDA_PREFIX to  `exists_conda_env` for a reason: https://github.com/Toblerity/Shapely/pull/843

@niklasarnitz what issue are you running into that you need to change this?

File: shapely/geos.py (hunk 1 above)
Comment (rburema) [reply shapely/geos.py:24]: I understand that the issue is larger than just us, but Cura runs 3.5 exclusively at the moment, so for us this isn't needed at all either.

File: shapely/geos.py (hunk 3 above)
Comment (melioristic) [reply shapely/geos.py:875]: > In the latest macOS (Big Sur), apple changed the way libraries are found and linked.
> In the current version of shapely, nothing works because of this:
> 
> ```
> Traceback (most recent call last):
>   File "<frozen importlib._bootstrap>", line 968, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 957, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 697, in exec_module
>   File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
>   File "/Users/ultimaker/build/4.6/build/inst/lib/python3.5/site-packages/cura/CuraApplication.py", line 51, in <module>
>   File "/Users/ultimaker/build/4.6/build/inst/lib/python3.5/site-packages/cura/Arranging/Arrange.py", line 7, in <module>
>   File "/Users/ultimaker/build/4.6/build/inst/lib/python3.5/site-packages/UM/Math/Polygon.py", line 10, in <module>
>   File "/Users/ultimaker/build/4.6/build/inst/lib/python3.5/site-packages/UM/Math/ShapelyUtil.py", line 6, in <module>
>   File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/geometry/__init__.py", line 4, in <module>
>   File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/geometry/base.py", line 17, in <module>
>   File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/coords.py", line 8, in <module>
>   File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/geos.py", line 113, in <module>
>   File "/Users/ultimaker/build/env/4.6/inst/lib/python3.5/site-packages/shapely/geos.py", line 56, in load_dll
> OSError: Could not find lib c or load any of its variants [].
> ```
> 
> The change in geos is only because it fixes the error in Cura (see this Issue: [#7970] ([Ultimaker/Cura#7970](https://github.com/Ultimaker/Cura/issues/7970)) over at Cura) and makes it work on Big Sur.
> I thought I would create a PR so you'll see it :)

Just to confirm, having the same issue. By when is this supposed to work or should I downgrade my mac.

Comment (rburema) [reply shapely/geos.py:875]: If you mean in Cura specifically, we've tried to implement a fix in our mac-build-environment that _should_ fix it in the 4.7.0 actual release (but not in the current beta!) Also if your specific issue is Cura, I'm sure the Shapely people'd appreciate it (not to mention us Cura devs) if you posted in the thread linked above, instead of here :-)

If you mean for Shapely in general, no idea! (We hope that someone --which might have to be us, though we lack some of the mac experience-- will look at this before we push out 4.8 so we don't have to keep a hack in our build system :-) )

Comment (melioristic) [reply shapely/geos.py:875]: I had this problem while I was using Cartopy. So in general I wanted to know about the shapely so that I can work with Cartopy.
Thanks for your quick reply. :))

**Comment by johnnync13** [comment]
Error with M1 - https://github.com/Toblerity/Shapely/issues/1080

**Comment by sgillies** [comment]
I'm closing this one. It's my understanding that Python 3.9.1 is the only Python version that supports macOS 11. And I have reports that the macosx Python wheel for Shapely 1.7.1 works with Big Sur.
